### PR TITLE
408 analytics per tenant

### PR DESCRIPTION
**Pull request by alishaevn**
# Story
- fixes: #408
- fixes: #409 
- related: https://github.com/samvera/hyrax/pull/6063
- related: #482
- related: #487 

# Expected Behavior Before Changes
- analytics were set through the `.env` or deploy.tmpl files
- one analytics account related to all tenants
- the feature is on by default

# Expected Behavior After Changes
- analytics info has to be set per tenant
- the feature is off by default

# Screenshots / Video

<details>
<summary>analytics by default</summary>

Screenshot 2023-05-17 at 1 09 51 PM
Screenshot 2023-05-17 at 1 09 04 PM
</details>

<details>
<summary>analytics enabled</summary>

Screenshot 2023-05-17 at 1 06 17 PM
Screenshot 2023-05-17 at 1 03 57 PM
</details>

### CODE REVIEW

File: app/services/hyrax/analytics/google.rb [hunk 1]
```diff
@@ -0,0 +1,214 @@
+# frozen_string_literal: true
+
+require 'oauth2'
+require 'signet/oauth_2/client'
+
+# OVERRIDE: Hyrax hyrax-v3.5.0 to make analytics a bit more dynamic
+# TODO(alishaevn): remove once https://github.com/samvera/hyrax/pull/6063 is available in our current hyrax version
```
Comment (alishaevn) [inline app/services/hyrax/analytics/google.rb:7]: for this reason, rob said to not make a decorator for this file.

File: config/analytics.yml [hunk 2]
```diff
@@ -6,6 +6,7 @@ analytics:
     analytics_id: <%= ENV['GOOGLE_ANALYTICS_ID'] %>
     app_name: <%= ENV['GOOGLE_OAUTH_APP_NAME'] %>
     app_version: <%= ENV['GOOGLE_OAUTH_APP_VERSION'] %>
-    privkey_path: <%= ENV['GOOGLE_OAUTH_PRIVATE_KEY_PATH'] %>
+    privkey_value: <%= ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'] %>
```
Comment (aprilrieger) [inline config/analytics.yml:9]: If you take a look at the worker, the way this works on a deployment we need the path to be setup and not the value.

Comment (alishaevn) [reply config/analytics.yml:9]: hmm. I believe I should delete that then since we won't have analytics related env variables. that shouldn't cause a problem right?

Comment (aprilrieger) [reply config/analytics.yml:9]: I am not sure, did you have issue navigating around in the dashboard after a dc down -v and a dc build?

I know I had an issue with it off main, and altered the worker script to add the `ENV['HYRAX_ANALYTICS']` to the if clause so the worker honors the flipper:

```yaml
if ENV['HYRAX_ANALYTICS'] && ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'] && !ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'].empty?
  %x{echo "$GOOGLE_OAUTH_PRIVATE_KEY_VALUE" | base64 -d > prod-cred.p12}
end
```

Comment (alishaevn) [reply config/analytics.yml:9]: I'll test this.

Comment (alishaevn) [reply config/analytics.yml:9]:
``` bash
docker compose down -v
docker compose build
docker compose up -d && docker compose exec web sh
  bundle exec rails s -b 0.0.0.0 -p 3000
```

after the above, the app still loaded for me. I'll do a build with no cache overnight to be doubly sure though.

Comment (alishaevn) [reply config/analytics.yml]: I made the changes in commit [b03fe08](https://github.com/scientist-softserv/palni-palci/pull/479/commits/b03fe08c2477d70b9daaae931ff7871118d75c7d), ran `dc down -v` and `dc build --no-cache`. that failed because of a wip file I had introduced. I removed that file and ran `dc build` again, which passed. starting up the app with `dc up` shows that the web and worker containers have started (in my docker dashboard). however, I see no logs for either in my terminal or in the dashboard. see below. I also only get the dory "http 502" error in my browser. I've confirmed dory is running though.

<details>
<summary>dashboard and browser</summary>

image

image
</details>

I don't understand why this is. the values I removed from the web and worker bin files are no longer used or referenced in this branch so it shouldn't cause a problem.

as info: I typically start up the app in a detached state, then manually start the server myself. (as noted in my above comment). so idk if the `dc up` command is also problematic on `main` for me.

File: .env [hunk 3]
```diff
@@ -47,16 +47,5 @@ HYKU_MULTITENANT=true
...
 
-##START## Enable Google Analytics
-# Uncomment to enable and configure Google Analytics, see README for instructions.
-HYRAX_ANALYTICS=true
-GOOGLE_ANALYTICS_ID=see-1pass
-GOOGLE_OAUTH_APP_NAME=palni-palci-demo
-GOOGLE_OAUTH_APP_VERSION=1.0
-GOOGLE_OAUTH_PRIVATE_KEY_SECRET=notasecret
-GOOGLE_OAUTH_PRIVATE_KEY_PATH=cred.p12
-GOOGLE_OAUTH_CLIENT_EMAIL=palni-palci-demo@palni-palci-demo.iam.gserviceaccount.com
```
Comment (alishaevn) [inline .env:58]: analytics are determined on a per tenant basis, so these are no longer necessary.

File: app/services/hyrax/analytics/google.rb [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+module Hyrax
+  module Analytics
+    module Google
+      extend ActiveSupport::Concern
+      # rubocop:disable Metrics/BlockLength
+      class_methods do
+        # Loads configuration options from config/analytics.yml for Hyrax apps only. You only need PRIVATE_KEY_PATH or
+        # PRIVATE_KEY_VALUE. VALUE takes precedence.
```
Comment (alishaevn) [inline app/services/hyrax/analytics/google.rb:16]: note that this is for hyrax only. and fix spelling errors.

File: app/services/hyrax/analytics/google.rb [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        # Expected structure:
+        # `analytics:`
+        # `  google:`
+        # `    app_name: <%= ENV['GOOGLE_OAUTH_APP_NAME']`
+        # `    app_version: <%= ENV['GOOGLE_OAUTH_APP_VERSION']`
+        # `    privkey_path: <%= ENV['GOOGLE_OAUTH_PRIVATE_KEY_PATH']`
+        # `    privkey_value: <%= ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE']`
+        # `    privkey_secret: <%= ENV['GOOGLE_OAUTH_PRIVATE_KEY_SECRET']`
+        # `    client_email: <%= ENV['GOOGLE_OAUTH_CLIENT_EMAIL']`
```
Comment (alishaevn) [inline app/services/hyrax/analytics/google.rb:25]: although I deleted the "analytics.yml" file for this app (as can happen with other hyku apps with this implementation), this file belongs to hyrax. so I left this comment.

File: app/services/hyrax/analytics/google.rb [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+            @config = config
+          end
+
+          # @return [Boolean] are all the required values present?
+          def valid?
+            config_keys = @config.keys
+            # OVERRIDE: Hyrax hyrax-v3.5.0 to require either setting
+            return false unless config_keys.include?('privkey_value') || config_keys.include?('privkey_path')
+
+            REQUIRED_KEYS.all? { |required| config_keys.include?(required) && @config[required].present? }
```
Comment (alishaevn) [inline app/services/hyrax/analytics/google.rb:64]: without `&& @config[required].present?`, this would return true simply because the keys existed on the config. although they were all empty strings.

File: config/analytics.yml (hunk 2 above)
Comment (alishaevn) [reply config/analytics.yml]: bringing up the app in a detached state and manually starting the server, works.

Comment (alishaevn) [reply config/analytics.yml]: april tested this locally and it worked for her so I deployed this branch to the staging env. it worked! no analytics are set by default, as expected.

image

File: app/services/hyrax/analytics/google.rb (hunk 6 above)
Comment (orangewolf) [reply app/services/hyrax/analytics/google.rb:64]: that makes sense, but that means we should change line 62

File: app/services/hyrax/analytics/google.rb [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+
+          def initialize(config)
+            @config = config
+          end
+
+          # @return [Boolean] are all the required values present?
+          def valid?
+            config_keys = @config.keys
+            # OVERRIDE: Hyrax hyrax-v3.5.0 to require either setting
+            return false unless config_keys.include?('privkey_value') || config_keys.include?('privkey_path')
```
Comment (orangewolf) [inline app/services/hyrax/analytics/google.rb:62]:
```suggestion
            return false unless @config['privkey_value'].present? || @config['privkey_path'].present?
```

File: app/services/hyrax/analytics/google.rb (hunk 6 above)
Comment (orangewolf) [inline app/services/hyrax/analytics/google.rb:64]:
```suggestion
            REQUIRED_KEYS.all? { |required| @config[required].present? }
```

Comment (orangewolf) [reply app/services/hyrax/analytics/google.rb:64]: actually I think the config_keys.include? is reduant. its either present or missing or blank and we don't care which of the later two it is.

File: bin/web [hunk 8]
```diff
@@ -1,6 +1,2 @@
 #!/usr/local/bin/ruby
-if ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'] && !ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'].empty?
```
Comment (orangewolf) [inline bin/web:2]: @alishaevn this needs to be a separate PR. You want to go to production with this still working and THEN fill in the tenants. once that is done, you can remove this as a different PR -> deploy step.

File: bin/worker [hunk 9]
```diff
@@ -1,8 +1,4 @@
 #!/usr/local/bin/ruby
-if ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'] && !ENV['GOOGLE_OAUTH_PRIVATE_KEY_VALUE'].empty?
```
Comment (orangewolf) [inline bin/worker:2]: @alishaevn this needs to be a separate PR. You want to go to production with this still working and THEN fill in the tenants. once that is done, you can remove this as a different PR -> deploy step.

File: config/analytics.yml [hunk 10]
```diff
@@ -1,11 +1,12 @@
 #
 # To integrate your app with Google Analytics, uncomment the lines below and add your API key information.
 #
-analytics:
```
Comment (orangewolf) [inline config/analytics.yml:4]: this template should still exists, we can just unset the variables it if we do not want tenants to fall back to a default. I'm not even sure we do want that?

File: config/browse_everything_providers.yml [hunk 11]
```diff
@@ -12,8 +12,8 @@ file_system:
 #   :client_id: YOUR_BOX_CLIENT_ID
 #   :client_secret: YOUR_BOX_CLIENT_SECRET
 # google_drive:
-#   :client_id: YOUR_GOOGLE_API_CLIENT_ID
```
Comment (orangewolf) [inline config/browse_everything_providers.yml:15]: this is incorrect. Google drive and Google analytics are two different things.

File: config/locales/simple_form.en.yml [hunk 12]
```diff
@@ -13,6 +13,13 @@ en:
           url: Fedora URL should not end with a slash
         name: A single or hyphenated name used for technical aspects of the repository (e.g., "acme" or "acme-library").
         gtm_id: The ID of your Google Tag Manager account
+        analytics_id: The ID of your Analytics account
+        analytics_oauth_app_name: The name of the application
```
Comment (orangewolf) [inline config/locales/simple_form.en.yml:17]:
```suggestion
        analytics_oauth_app_name: The name of the Google application in the Google API console
```

File: config/locales/simple_form.en.yml [hunk 13]
```diff
@@ -13,6 +13,13 @@ en:
...
+        analytics_oauth_app_version: The version of application
```
Comment (orangewolf) [inline config/locales/simple_form.en.yml:18]:
```suggestion
        analytics_oauth_app_version: The version of Google application in the Google API console
```

File: config/locales/simple_form.en.yml [hunk 14]
```diff
@@ -13,6 +13,13 @@ en:
...
+        analytics_oauth_private_key_secret: The secret provided when you created the key
```
Comment (orangewolf) [inline config/locales/simple_form.en.yml:19]: can you review these last 3 with the idea of "by who and where?" added in? there is a lot implied by these as terse as they are.

File: ops/staging-deploy.tmpl.yaml [hunk 15]
```diff
@@ -86,19 +86,19 @@ extraEnvVars: &envVars
     value: http://fcrepo.staging-fcrepo.svc.cluster.local:8080/rest
   - name: GOOGLE_FONTS_KEY
     value: $GOOGLE_FONTS_KEY
-  # - name: GOOGLE_ANALYTICS_ID
-  #   value: $GOOGLE_ANALYTICS_ID  # Set in GitHub Environment Variables Section
-  # - name: GOOGLE_OAUTH_APP_NAME
+  # - name: ANALYTICS_ID
```
Comment (orangewolf) [inline ops/staging-deploy.tmpl.yaml:89]: I don't feel like I understand this change?

**Review by orangewolf — CHANGES_REQUESTED** [review]

File: ops/staging-deploy.tmpl.yaml (hunk 15 above)
Comment (alishaevn) [reply ops/staging-deploy.tmpl.yaml:89]: @orangewolf the idea was that since this code is going back into hyku, to not limit the person to using google analytics. even though that is what pals is using. so instead of prefacing everything with "google", we leave it more open.

the change in this file specifically is to match what we have in "account_settings.rb" and everywhere else now.

File: bin/worker (hunk 9 above)
Comment (alishaevn) [reply bin/worker:2]: ok, I'll move this to #482.

File: bin/web (hunk 8 above)
Comment (alishaevn) [reply bin/web:2]: ok, I'll move this to #482.

File: config/analytics.yml (hunk 10 above)
Comment (alishaevn) [reply config/analytics.yml:4]: you mean we don't want the fallback? I agree, and had deleted this file actually. but it fails at [self.load_from_yaml](https://github.com/scientist-softserv/palni-palci/blob/54a7b255a1d71ac55da0e2ba624d0d264fdcf2b0/app/services/hyrax/analytics/google.rb#L33-L34) since the file doesn't exist. I saw that we have [analytics.yml commented out](https://github.com/samvera/hyku/blob/main/config/analytics.yml) in base hyku, and do not have a "google.rb" file/decorator, so I commented the file out too instead of trying to get around the failure.

would you prefer I try to get around the error?

File: config/locales/simple_form.en.yml (hunk 14 above)
Comment (alishaevn) [reply config/locales/simple_form.en.yml:19]: > by who

I pulled these from what was in the readme and left them vague since this work is going back to hyku, which can also use matomo. and potentially other applications in the future? 

> and where

I wasn't sure where these things could be found for matomo. 
<hr>

would you prefer I explicitly state google and matomo for now? and look up where these things can be found for matomo?

File: config/locales/simple_form.en.yml (hunk 13 above)
Comment (alishaevn) [reply config/locales/simple_form.en.yml:18]: ref: https://github.com/scientist-softserv/palni-palci/pull/479#discussion_r1198129112

File: config/locales/simple_form.en.yml (hunk 12 above)
Comment (alishaevn) [reply config/locales/simple_form.en.yml:17]: ref: https://github.com/scientist-softserv/palni-palci/pull/479#discussion_r1198129112

File: ops/staging-deploy.tmpl.yaml (hunk 15 above)
Comment (alishaevn) [reply ops/staging-deploy.tmpl.yaml:89]: although they will all get removed too on #482

File: bin/worker (hunk 9 above)
Comment (alishaevn) [reply bin/worker]: moved!

File: bin/web (hunk 8 above)
Comment (alishaevn) [reply bin/web]: moved!

File: config/browse_everything_providers.yml (hunk 11 above)
Comment (alishaevn) [reply config/browse_everything_providers.yml]: reverted!

File: bin/worker (hunk 9 above)
Comment (alishaevn) [reply bin/worker]: moved this to #487 instead.

**Review by orangewolf — APPROVED** [review]

File: config/locales/simple_form.en.yml (hunk 14 above)
Comment (alishaevn) [reply config/locales/simple_form.en.yml:19]: ref: #492
